xds/internal/balancer/outlierdetection/balancer.go

```diff
@@ -105,3 +132,760 @@ func (bb) ParseConfig(s json.RawMessage) (serviceconfig.LoadBalancingConfig, err
+
+	if info.Err == nil {
+		atomic.AddInt64(&ab.numSuccesses, 1)
+	} else {
+		atomic.AddInt64(&ab.numFailures, 1)
+	}
+	atomic.AddInt64(&ab.requestVolume, 1)
+}
+
+func (b *outlierDetectionBalancer) UpdateState(s balancer.State) {
+	b.pickerUpdateCh.Put(s)
+
+}
+
+func (b *outlierDetectionBalancer) NewSubConn(addrs []resolver.Address, opts balancer.NewSubConnOptions) (balancer.SubConn, error) {
+	// "When the child policy asks for a subchannel, the outlier_detection will
+	// wrap the subchannel with a wrapper." - A50
+	sc, err := b.cc.NewSubConn(addrs, opts)
+	if err != nil {
+		return nil, err
+	}
+	scw := &subConnWrapper{
+		SubConn:    sc,
+		addresses:  addrs,
+		scUpdateCh: b.scUpdateCh,
+	}
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	b.scWrappers[sc] = scw
+	if len(addrs) != 1 {
+		return scw, nil
+	}
+	obj, ok := b.addrs[addrs[0].Addr]
+	if !ok {
+		return scw, nil
+	}
+	obj.sws = append(obj.sws, scw)
+	atomic.StorePointer(&scw.addressInfo, unsafe.Pointer(obj))
+
+	// "If that address is currently ejected, that subchannel wrapper's eject
+	// method will be called." - A50
+	if !obj.latestEjectionTimestamp.IsZero() {
+		scw.eject()
+	}
+	return scw, nil
+}
+
+func (b *outlierDetectionBalancer) RemoveSubConn(sc balancer.SubConn) {
+	scw, ok := sc.(*subConnWrapper)
+	if !ok { // Shouldn't happen
+		return
+	}
+	// Remove the wrapped SubConn from the parent Client Conn. We don't remove
+	// from map entry until we get a Shutdown state for the SubConn, as we need
+	// that data to forward that state down.
+	b.cc.RemoveSubConn(scw.SubConn)
+}
+
+// appendIfPresent appends the scw to the address, if the address is present in
+// the Outlier Detection balancers address map. Returns nil if not present, and
+// the map entry if present.
+func (b *outlierDetectionBalancer) appendIfPresent(addr string, scw *subConnWrapper) *addressInfo {
+	obj, ok := b.addrs[addr]
+	if !ok {
+		return nil
+	}
+
+	obj.sws = append(obj.sws, scw)
+	atomic.StorePointer(&scw.addressInfo, unsafe.Pointer(obj))
+	return obj
+}
+
+// removeSubConnFromAddressesMapEntry removes the scw from it's map entry if
+// present.
+func (b *outlierDetectionBalancer) removeSubConnFromAddressesMapEntry(scw *subConnWrapper) {
+	obj := (*addressInfo)(atomic.LoadPointer(&scw.addressInfo))
+	if obj == nil {
+		return
+	}
+	for i, sw := range obj.sws {
+		if scw == sw {
+			obj.sws = append(obj.sws[:i], obj.sws[i+1:]...)
+			break
+		}
+	}
+}
+
+func (b *outlierDetectionBalancer) UpdateAddresses(sc balancer.SubConn, addrs []resolver.Address) {
+	scw, ok := sc.(*subConnWrapper)
+	if !ok {
+		// Return, shouldn't happen if passed up scw
+		return
+	}
+
+	b.cc.UpdateAddresses(scw.SubConn, addrs)
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	// Note that 0 addresses is a valid update/state for a SubConn to be in.
+	// This is correctly handled by this algorithm (handled as part of a non singular
+	// old address/new address).
+	if len(scw.addresses) == 1 {
```

Comment:
This would probably be a little clearer as:

```go
if len(scw.addresses == 1) && len(addrs) == 1 {
	// single address to single address
} else if len(scw.addresses == 1) {
	// single address to multiple/no addresses
} else if len(addrs) == 1 {
	// multiple address to single address
} // else multiple/no addresses to multiple/no addresses; ignore
```
or:
```go
switch {
case len(scw.addresses == 1) && len(addrs) == 1:
case len(scw.addresses == 1):
case len(addrs) == 1:
}
```